src/python/pants/core/util_rules/environments.py

```diff
@@ -363,5 +314,109 @@ def extract_process_config_from_environment(tgt: EnvironmentTarget) -> ProcessCo
     return ProcessConfigFromEnvironment(docker_image=docker_image)
 
 
+class EnvironmentSensitiveOptionFieldMixin:
+    subsystem: ClassVar[type[Subsystem]]
+    option_name: ClassVar[str]
+
+
+# Maps between non-list option value types and corresponding fields
+_SIMPLE_OPTIONS: dict[type, type[Field]] = {
+    str: StringField,
+}
+
+# Maps between the member types for list options. Each element is the
+# field type, and the `value` type for the field.
+_LIST_OPTIONS: dict[type, type[Field]] = {
+    str: StringSequenceField,
+}
+
+
+@memoized
+def add_option_fields_for(subsystem: type[Subsystem]) -> Iterable[UnionRule]:
+    """Register environment fields for the `environment_sensitive` options of `subsystem`
+
+    This should be called in the `rules()` method in the file where `subsystem` is defined. It will
+    register the relevant fields under the `local_environment` and `docker_environment` targets.
+    """
+    field_rules: set[UnionRule] = set()
+
+    for option_attrname, option in _collect_options_info_extended(subsystem):
+        if option.flag_options["environment_sensitive"]:
+            field_rules.update(_add_option_field_for(subsystem, option, option_attrname))
```

Comment:
Rather than using the `attr` that the `OptionInfo` is registered under, this should probably be using the flag name, right? The `attr` is helpfully already snake-cased, but there is no requirement that the attr name matches the flag name... it's just a default flag name.

Reply:
To be clear, the field name will use the flag name. The attr name is necessary so that it's possible to reconcile the field with the relevant attr when you're retrieving the value from the `Subsystem`.